<h1><b>Nama:     Muhamad Al Kausar Ramadhan</b></h1>
<h1><b>Kelas:    TI 3A</b></h1>
<h1><b>Absen:    15</b></h1>
<h1><b>Nim:      2041720193</b></h1>

## Soal
1. Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)
2. Tentukan data validasi! (10 poin)
3. Tentukan secara tepat kolom yang menjadi "label" (20 poin)
4. Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)
5. Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya. (40 poin)

In [54]:
#memanggil library
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression


In [55]:
#df digunakan sebagai variable untuk memanggil file excel yang bernama dataset_quiz yang ada di folder data
df = pd.read_excel('data/dataset_quiz.xlsx')
df.head(100)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0     7206.0       7214.0       6211.0       8602.0       7801.0          NaN   
1     7214.0       7109.0       8402.0       7903.0       9202.0          NaN   
2     5911.0       8201.0       8701.0       9101.0       9003.0          NaN   
3     8402.0       7702.0       8602.0       8903.0       8903.0          NaN   
4     8201.0       8305.0       5911.0       5903.0       8402.0          NaN   
..       ...          ...          ...          ...          ...          ...   
80    8803.0       8804.0       9102.0       8005.0       9303.0          NaN   
81    6211.0       8901.0       9003.0       9303.0       9007.0          NaN   
82    8601.0       8903.0       8703.0          NaN          NaN          NaN   
83    8703.0       8601.0       9202.0       9102.0       6204.0          NaN   
84    8102.0       7109.0       5903.0       7206.0       8903.0          NaN   

   keterlibatan_vltr  jml_vltr_terlibat  tahun    skema luaran_hki  \
0              TIDAK                  0   2019  REGULER  TIDAK ADA   
1              TIDAK                  0   2019  REGULER  TIDAK ADA   
2              TIDAK                  0   2019  REGULER  TIDAK ADA   
3              TIDAK                  0   2019  REGULER  TIDAK ADA   
4              TIDAK                  0   2019  REGULER  TIDAK ADA   
..               ...                ...    ...      ...        ...   
80                YA                  3   2022  REGULER  TIDAK ADA   
81                YA                  3   2022  REGULER  TIDAK ADA   
82                YA                  3   2022  REGULER  TIDAK ADA   
83                YA                  3   2022  REGULER  TIDAK ADA   
84             TIDAK                  0   2022  REGULER  TIDAK ADA   

   jenis_luaran_hki potensi_hki luaran_lain jenis_luaran_lain  \
0         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
1         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
2         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
3         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
4         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
..              ...         ...         ...               ...   
80        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
81        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
82        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
83        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
84        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   

   status_luaran_lain  
0           TIDAK ADA  
1           TIDAK ADA  
2           TIDAK ADA  
3           TIDAK ADA  
4           TIDAK ADA  
..                ...  
80          TIDAK ADA  
81          TIDAK ADA  
82          TIDAK ADA  
83          TIDAK ADA  
84          TIDAK ADA  

[85 rows x 16 columns]

In [56]:
#untuk mengecek apakah ada data yang kosong
df.isnull().sum()

id_ketua               3
id_anggota1            5
id_anggota2            4
id_anggota3            3
id_anggota4           13
id_anggota5           84
keterlibatan_vltr      0
jml_vltr_terlibat      0
tahun                  0
skema                  0
luaran_hki             0
jenis_luaran_hki       0
potensi_hki            0
luaran_lain            0
jenis_luaran_lain      0
status_luaran_lain     0
dtype: int64

In [57]:
#new df digunakan untuk menampung data yang dimana column id_anggota 5 sudah di drop, 
# lalu variable new df dipanggil untuk di cek apakah column id_anggota 5 berhasil dihapus
new_df = df.drop('id_anggota5', axis=1)
new_df.head()

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  \
0    7206.0       7214.0       6211.0       8602.0       7801.0   
1    7214.0       7109.0       8402.0       7903.0       9202.0   
2    5911.0       8201.0       8701.0       9101.0       9003.0   
3    8402.0       7702.0       8602.0       8903.0       8903.0   
4    8201.0       8305.0       5911.0       5903.0       8402.0   

  keterlibatan_vltr  jml_vltr_terlibat  tahun    skema luaran_hki  \
0             TIDAK                  0   2019  REGULER  TIDAK ADA   
1             TIDAK                  0   2019  REGULER  TIDAK ADA   
2             TIDAK                  0   2019  REGULER  TIDAK ADA   
3             TIDAK                  0   2019  REGULER  TIDAK ADA   
4             TIDAK                  0   2019  REGULER  TIDAK ADA   

  jenis_luaran_hki potensi_hki luaran_lain jenis_luaran_lain  \
0        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
1        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
2        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
3        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
4        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   

  status_luaran_lain  
0          TIDAK ADA  
1          TIDAK ADA  
2          TIDAK ADA  
3          TIDAK ADA  
4          TIDAK ADA

In [58]:
#disini saya asumsikan bahwa column id_ketua dan id_anggota 1 - 4 tidak diperlukan
#sehingga saya melakukan pembuatan variable baru yaitu new_df2 dimana hanya mengambil kolumn keterlibatan_vltr - status_luaran_lain
new_df2 = new_df.iloc[:, 5:]

#lalu saya melakukan encoding menggunakan perulangan for untuk menguba isi column yang isinya data kategori menjadi angka
from sklearn.preprocessing import LabelEncoder
labelencoder=LabelEncoder()
for column in new_df2.columns:
    new_df2[column] = labelencoder.fit_transform(new_df2[column])

In [59]:
new_df2.head()

keterlibatan_vltr  jml_vltr_terlibat  tahun  skema  luaran_hki  \
0                  0                  0      0      1           1   
1                  0                  0      0      1           1   
2                  0                  0      0      1           1   
3                  0                  0      0      1           1   
4                  0                  0      0      1           1   

   jenis_luaran_hki  potensi_hki  luaran_lain  jenis_luaran_lain  \
0                 1            2            0                  3   
1                 1            2            0                  3   
2                 1            2            0                  3   
3                 1            2            0                  3   
4                 1            2            0                  3   

   status_luaran_lain  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1

In [60]:
#lalu variable new_df2 saya gunakan dropna untuk menghapus isi - isi row yang kosong
new_df2 = new_df2.dropna()
#lalu dilakukan pengecekan apakah ada data yang kosong
new_df2.isnull().sum()

keterlibatan_vltr     0
jml_vltr_terlibat     0
tahun                 0
skema                 0
luaran_hki            0
jenis_luaran_hki      0
potensi_hki           0
luaran_lain           0
jenis_luaran_lain     0
status_luaran_lain    0
dtype: int64

In [61]:
# lalu baris ini untuk mengecek ada berapa banyak data setelah dilakukan penghapusan menggunakan dropna
print(new_df2['keterlibatan_vltr'].value_counts())

0    45
1    40
Name: keterlibatan_vltr, dtype: int64


<h4>Jawaban Nomor 3 </h4>
<p>Tentukan secara tepat kolom yang menjadi "label" (20 poin)</p>
<b><p>Kolumn keterlibatan vltr menjadi "label"</p></b>

In [62]:
#X mengambil data dari kolumn 1 (jml_vltr_terlibat) - kolumn terakhir (status_luaran_lain)
X = new_df2.iloc[:, 1:]
#y mengambil data dari kolumn 0 (keterlibatan_vtr)
y = new_df2.iloc[:, 0]
X.head()

jml_vltr_terlibat  tahun  skema  luaran_hki  jenis_luaran_hki  potensi_hki  \
0                  0      0      1           1                 1            2   
1                  0      0      1           1                 1            2   
2                  0      0      1           1                 1            2   
3                  0      0      1           1                 1            2   
4                  0      0      1           1                 1            2   

   luaran_lain  jenis_luaran_lain  status_luaran_lain  
0            0                  3                   1  
1            0                  3                   1  
2            0                  3                   1  
3            0                  3                   1  
4            0                  3                   1

<h4>Jawaban Nomor 1 </h4>
<p>Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)</p>

In [63]:
#memanggil librar train_test_split
from sklearn.model_selection import train_test_split
# Split data training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

<h4>Jawaban Nomor 2 </h4>
<p>Tentukan data validasi! (10 poin)</p>

In [64]:
# Split data training dan dan lainnya
# data lainnya, akan kita split lagi menjadi validasi dan testing.
# Rasio yang akan kita gunakan adalah 8:1:1
df_train, df_unseen = train_test_split(new_df2, test_size=0.2, random_state=0)

# Split lagi antara validasi dan testing
df_val, df_test = train_test_split(df_unseen, test_size=0.5, random_state=0)

# Cek masing-masing ukuran data

print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_train.shape[0]}')
print(f'Jumlah data val: {df_val.shape[0]}')
print(f'Jumlah data test: {df_test.shape[0]}')

# Cek rasio tiap label
print('=========')
print(f'Jumlah label data asli:\n{new_df2.keterlibatan_vltr	.value_counts()}')
print(f'Jumlah label data train:\n{df_train.keterlibatan_vltr	.value_counts()}')
print(f'Jumlah label data val:\n{df_val.keterlibatan_vltr	.value_counts()}')
print(f'Jumlah label data test:\n{df_test.keterlibatan_vltr	.value_counts()}')

Jumlah data asli: 85
Jumlah data train: 68
Jumlah data val: 8
Jumlah data test: 9
Jumlah label data asli:
0    45
1    40
Name: keterlibatan_vltr, dtype: int64
Jumlah label data train:
0    36
1    32
Name: keterlibatan_vltr, dtype: int64
Jumlah label data val:
1    5
0    3
Name: keterlibatan_vltr, dtype: int64
Jumlah label data test:
0    6
1    3
Name: keterlibatan_vltr, dtype: int64


<h4>Jawaban Nomor 4 </h4>
<p>Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)</p>

<b>DecisionTreeClassifier</b>

In [65]:
from sklearn.tree import DecisionTreeClassifier

# Instantiate DecisionTreeClassifier 'dt' dengan kedalaman maksimum 6
# Secara default, DecisionTreeClassifier dari scikit-learn akan menggunakan nilai "Gini" untuk kriteria
# Terdapat beberapa "hyperparamater" yang dapat digunakan. Silahka baca dokumentasi
dt = DecisionTreeClassifier()

# menyesuaikan dt ke set training
dt.fit(X_train, y_train)

DecisionTreeClassifier()

In [66]:
from sklearn.metrics import accuracy_score

# Memprediksi test set labels
y_pred = dt.predict(X_test)

#  menghitung set accuracy
acc = accuracy_score(y_test, y_pred)
print(f"Test set accuracy: {acc}")

Test set accuracy: 1.0


<b>Naive Bayes - Multinomial NB</b>

In [67]:
from sklearn.naive_bayes import MultinomialNB # class untuk model MultinomialNB

# Inisiasi obyek MultinomialNB
mnb = MultinomialNB()

# Kita dapat langsung menggunakan fitur X dan label y
# hasil dari proses pembuatan data dummy

# Split data training dan testing
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=1)

# Fit model
# Label y harus dalam bentuk 1D atau (n_samples,)
mnb.fit(X_train, y_train)

# Prediksi dengan data training
y_train_pred = mnb.predict(X_train)

# Evaluasi akurasi training
acc_train = accuracy_score(y_train, y_train_pred)

# Prediksi test data
y_test_pred = mnb.predict(X_test)

# Evaluasi model dengan metric akurasi
acc_test = accuracy_score(y_test, y_test_pred)

# Print hasil evaluasi
print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')

Hasil akurasi data train: 0.9852941176470589
Hasil akurasi data test: 1.0


<h4>Jawaban Nomor 5 </h4>
<p>Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya. (40 poin)</p>

<b>Random Search</b>

In [68]:
#Memanggil RepeatedKFold
from sklearn.model_selection import RepeatedKFold
# Inisiasi obyek RepeatedKFold dengan split 4, repeats 3, dan random_state = 42
rkf = RepeatedKFold(n_splits=4, n_repeats=3, random_state=42)

In [69]:
#Memanggil RandomizedSearchCV
from sklearn.model_selection import RandomizedSearchCV

# Inisiasi obyek DecisionTreeClassifier
dtc = DecisionTreeClassifier()

#set parameter dengan criterion : gini, entropy, log_loss (wajib) dengan max depth di antara 5 - 11
params = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11))
}

# Inisiasi obyek RandomizedSearchCV dengan estimator yang dipakai adalah dtc
# value parameter nya adalah params, cv adalah rkf sebagai validasi silang K-fold
random = RandomizedSearchCV(dtc, param_distributions=params, cv=rkf)
%timeit random.fit(X, y)

#mengambil nilai evaluasi dari X dan y
score = random.score(X,y)

#cetak hasilnya
print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {random.best_params_}')

c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 593, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 898, in fit
    super().fit(
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 347, in fit
    criterion = CRITERIA_CLF[self.criterion](self.n_outputs_,
KeyError: 'log_loss'

  warnings.warn("Estimator fit failed. The score on this train-test"
c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceb

587 ms ± 8.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'max_depth': 5, 'criterion': 'entropy'}


c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 593, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 898, in fit
    super().fit(
  File "c:\Users\user\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 347, in fit
    criterion = CRITERIA_CLF[self.criterion](self.n_outputs_,
KeyError: 'log_loss'

  warnings.warn("Estimator fit failed. The score on this train-test"
c:\Users\user\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceb

<b>Grid Search</b>

In [70]:
# Memanggil GridSearchCV dan SVC
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

# Inisiasi objek SVC
svc = SVC()

#mendefinisikan range parameter
params = {
    'kernel': ['rbf', 'linear', 'poly'],
    'C': [3],
    'degree': [1]
}

# Inisiasi obyek GridSearchCV estimator yang dipakai adalah SVC, 
# value parameter nya adalah params, cv adalah rkf sebagai validasi silang K-fold
grid = GridSearchCV(svc, param_grid=params, cv=rkf, n_jobs=-1)

#proses pengambilan nilai scorenya
%timeit grid.fit(X, y)
score = grid.score(X,y)

#mendapatkan hasilnya
print(f'Hasil evaluasi Grid Search: {score}')
print(f'Konfigurasi hyperparameter Grid Search: {grid.best_params_}')

125 ms ± 14.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi Grid Search: 0.9882352941176471
Konfigurasi hyperparameter Grid Search: {'C': 3, 'degree': 1, 'kernel': 'rbf'}


<h3>KESIMPULAN</h3>
<p>Hasil Akurasi Dengan Random Search Lebih Tinggi Daripada Grid Search</p>
<li>Random Search : 1.00</li>
<li>Grid Search : 0.988</li>
<br>
<b><p>Bisa Disimpulkan Bahwa Hasil Grid Search Lebih Baik Dibandingkan Dengan Random Search</p></b>